handler/src/main/java/io/netty/handler/pcap/PCAPWriteHandler.java

```diff
@@ -0,0 +1,410 @@
+            } catch (IOException ex) {
+                ctx.fireExceptionCaught(ex);
+            } finally {
+                byteBuf.release();
+            }
+
+        } else {
+            this.pCapWriter = new PCapWriter(this.outputStream);
+        }
+
+        if (ctx.channel() instanceof SocketChannel) {
+            if (ctx.channel() instanceof ServerChannel) {
+                srcAddr = (InetSocketAddress) ctx.channel().remoteAddress();
+                dstAddr = (InetSocketAddress) ctx.channel().localAddress();
+            } else {
+                srcAddr = (InetSocketAddress) ctx.channel().localAddress();
+                dstAddr = (InetSocketAddress) ctx.channel().remoteAddress();
+            }
+        }
+
+        if (ctx.channel() instanceof SocketChannel) {
+            logger.debug("Starting Fake TCP 3-Way Handshake");
+
+            ByteBuf tcpBuf = ctx.alloc().buffer();
+
+            try {
+                // Write SYN with Normal Source and Destination Address
+                TCPPacket.writePacket(tcpBuf, null, 0, 0, srcAddr.getPort(), dstAddr.getPort(), TCPPacket.TCPFlag.SYN);
+                completeTCPWrite(srcAddr, dstAddr, tcpBuf, ctx.alloc(), ctx);
+
+                // Write SYN+ACK with Reversed Source and Destination Address
+                TCPPacket.writePacket(tcpBuf, null, 0, 1, dstAddr.getPort(), srcAddr.getPort(), TCPPacket.TCPFlag.SYN,
+                        TCPPacket.TCPFlag.ACK);
+                completeTCPWrite(dstAddr, srcAddr, tcpBuf, ctx.alloc(), ctx);
+
+                // Write ACK with Normal Source and Destination Address
+                TCPPacket.writePacket(tcpBuf, null, 1, 1, srcAddr.getPort(), dstAddr.getPort(), TCPPacket.TCPFlag.ACK);
+                completeTCPWrite(srcAddr, dstAddr, tcpBuf, ctx.alloc(), ctx);
+            } finally {
+                tcpBuf.release();
+            }
+
+            logger.debug("Finished Fake TCP 3-Way Handshake");
+        }
+
+        super.channelActive(ctx);
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        writePacket(ctx, msg, false);
+        super.channelRead(ctx, msg);
+    }
+
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
+        writePacket(ctx, msg, true);
+        super.write(ctx, msg, promise);
+    }
+
+    private void writePacket(ChannelHandlerContext ctx, Object msg, boolean isWriteOperation) {
+        if (msg instanceof ByteBuf) {
+
+            // If bytes are 0 and `captureZeroByte` is false, we won't capture this.
+            if (((ByteBuf) msg).readableBytes() == 0 && !captureZeroByte) {
+                logger.debug("Discarding Zero Byte TCP Packet. isWriteOperation {}", isWriteOperation);
+                return;
+            }
+
+            ByteBuf packet = ((ByteBuf) msg).duplicate();
+            ByteBuf tcpBuf = ctx.alloc().buffer();
```

Comment:
this is not really correct... Even with UDP we may see `ByteBuf` messages if the channel is connected. I think you should use `ctx.channel() instanceof DatagramChannel` to decide if its UDP or not

Reply:
Handling the` EmbeddedChannel` with this approach will be an issue then for Unit Tests. I'm replacing the `EmbeddedChannel` with `Bootstrap` client in Test Case.